File: src/detect11/src/yolo_detect_node.cpp
#include <chrono>
#include <filesystem>
#include <iostream>

// ROS2 Includes
#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/image.hpp"
#include "cv_bridge/cv_bridge.h"
#include "message_filters/subscriber.h"
#include "message_filters/synchronizer.h"
#include "message_filters/sync_policies/approximate_time.h"
#include "ament_index_cpp/get_package_share_directory.hpp" 

// Custom Message
#include "object3d_msgs/msg/object3_d.hpp"
#include "object3d_msgs/msg/object3_d_array.hpp"

// Include our new detector library
#include "detect11/cpu_detect.hpp"

using namespace std::chrono_literals;

class YoloDetectNode : public rclcpp::Node {
public:
    YoloDetectNode() : Node("yolo_detect_node") {
        // 1. 声明参数
        this->declare_parameter("camera.fx", 905.5593);
        this->declare_parameter("camera.fy", 905.5208);
        this->declare_parameter("camera.cx", 663.4498);
        this->declare_parameter("camera.cy", 366.7621);
        this->declare_parameter("conf_thres", 0.50);
        this->declare_parameter("show_image", true);
        this->declare_parameter("model_filename", "yolo11n.onnx"); 

        fx_ = this->get_parameter("camera.fx").as_double();
        fy_ = this->get_parameter("camera.fy").as_double();
        cx_ = this->get_parameter("camera.cx").as_double();
        cy_ = this->get_parameter("camera.cy").as_double();
        conf_thres_ = this->get_parameter("conf_thres").as_double();
        show_image_ = this->get_parameter("show_image").as_bool();
        
        // 2. 加载模型逻辑
        std::string model_filename = this->get_parameter("model_filename").as_string();
        std::string final_model_path;
        namespace fs = std::filesystem;
        fs::path p(model_filename);

        if (p.is_absolute()) {
            final_model_path = model_filename;
        } else {
            try {
                std::string package_share_directory = ament_index_cpp::get_package_share_directory("detect11");
                final_model_path = (fs::path(package_share_directory) / "models" / model_filename).string();
            } catch (const std::exception& e) {
                RCLCPP_ERROR(this->get_logger(), "Package error: %s", e.what());
                return;
            }
        }

        if (!fs::exists(final_model_path)) {
            RCLCPP_ERROR(this->get_logger(), "Model not found: %s", final_model_path.c_str());
            return;
        }

        RCLCPP_INFO(this->get_logger(), "Initializing Yolo11Detector with: %s", final_model_path.c_str());
        
        try {
            // 初始化探测器类
            detector_ = std::make_unique<Yolo11Detector>(final_model_path, conf_thres_);
        } catch (const std::exception& e) {
            RCLCPP_ERROR(this->get_logger(), "Detector init failed: %s", e.what());
            return;
        }

        // 3. 订阅与发布
        rmw_qos_profile_t qos_profile = rmw_qos_profile_sensor_data;
        auto qos = rclcpp::QoS(rclcpp::QoSInitialization::from_rmw(qos_profile), qos_profile);

        color_sub_.subscribe(this, "/camera/realsense_d435i/color/image_raw", qos.get_rmw_qos_profile());
        depth_sub_.subscribe(this, "/camera/realsense_d435i/aligned_depth_to_color/image_raw", qos.get_rmw_qos_profile());

        sync_ = std::make_shared<message_filters::Synchronizer<SyncPolicy>>(SyncPolicy(10), color_sub_, depth_sub_);
        sync_->registerCallback(&YoloDetectNode::sync_callback, this);
        sync_->setMaxIntervalDuration(rclcpp::Duration::from_seconds(0.1));

        publisher_ = this->create_publisher<object3d_msgs::msg::Object3DArray>("target_points_array", qos);
        
        param_callback_handle_ = this->add_on_set_parameters_callback(
            std::bind(&YoloDetectNode::parametersCallback, this, std::placeholders::_1));

        RCLCPP_INFO(this->get_logger(), "Node Ready.");
    }

private:
    rcl_interfaces::msg::SetParametersResult parametersCallback(const std::vector<rclcpp::Parameter> &parameters) {
        rcl_interfaces::msg::SetParametersResult result;
        result.successful = true;
        for (const auto &param : parameters) {
            if (param.get_name() == "show_image" && param.get_type() == rclcpp::ParameterType::PARAMETER_BOOL) {
                show_image_ = param.as_bool();
                RCLCPP_INFO(this->get_logger(), "Update show_image: %s", show_image_ ? "true" : "false");
            }
        }
        return result;
    }

    float get_robust_depth(const cv::Mat& depth_img, int cx, int cy) {
        if (cx < 0 || cx >= depth_img.cols || cy < 0 || cy >= depth_img.rows) return -1.0;
        int x_min = std::max(0, cx - 2);
        int x_max = std::min(depth_img.cols, cx + 3);
        int y_min = std::max(0, cy - 2);
        int y_max = std::min(depth_img.rows, cy + 3);

        std::vector<unsigned short> valid_depths;
        valid_depths.reserve(25);
        for (int y = y_min; y < y_max; ++y) {
            for (int x = x_min; x < x_max; ++x) {
                unsigned short d = depth_img.at<unsigned short>(y, x);
                if (d > 0) valid_depths.push_back(d);
            }
        }
        if (valid_depths.empty()) return -1.0;
        size_t n = valid_depths.size() / 2;
        std::nth_element(valid_depths.begin(), valid_depths.begin() + n, valid_depths.end());
        return (float)valid_depths[n];
    }

    void sync_callback(const sensor_msgs::msg::Image::ConstSharedPtr& msg_color, 
                       const sensor_msgs::msg::Image::ConstSharedPtr& msg_depth) {
        // FPS
        auto now = std::chrono::steady_clock::now();
        frame_count_++;
        double elapsed = std::chrono::duration_cast<std::chrono::duration<double>>(now - start_time_).count();
        if (elapsed >= 1.0) {
            fps_ = frame_count_ / elapsed;
            frame_count_ = 0;
            start_time_ = now;
        }

        cv::Mat color_img, depth_img;
        try {
            color_img = cv_bridge::toCvShare(msg_color, "bgr8")->image;
            depth_img = cv_bridge::toCvShare(msg_depth, "16UC1")->image;
        } catch (cv_bridge::Exception& e) {
            RCLCPP_ERROR(this->get_logger(), "cv_bridge error: %s", e.what());
            return;
        }

        if (show_image_) {
            cv::putText(color_img, "FPS: " + std::to_string((int)fps_), cv::Point(10, 30), 
                        cv::FONT_HERSHEY_SIMPLEX, 1.0, cv::Scalar(0, 255, 0), 2);
        }

        // === 调用分离出去的推理引擎 ===
        auto detections = detector_->detect(color_img, show_image_);
        // ============================

        // 显示逻辑
        if (show_image_) {
            const std::string win_name = "YOLOv11 Detection";
            cv::namedWindow(win_name, cv::WINDOW_NORMAL);
            cv::setWindowProperty(win_name, cv::WND_PROP_FULLSCREEN, cv::WINDOW_FULLSCREEN);
            cv::imshow(win_name, color_img);
            cv::waitKey(1);
        } else {
            try {
                if (cv::getWindowProperty("YOLOv11 Detection", cv::WND_PROP_VISIBLE) >= 0) {
                    cv::destroyWindow("YOLOv11 Detection");
                    cv::waitKey(1);
                }
            } catch (...) {}
        }

        if (publisher_->get_subscription_count() == 0 || detections.empty()) return;

        object3d_msgs::msg::Object3DArray array_msg;
        array_msg.header = msg_color->header;

        for (const auto& det : detections) {
            float d = get_robust_depth(depth_img, det.center.x, det.center.y);
            if (d <= 0) continue;

            float Z = d / 1000.0f;
            float X = (det.center.x - cx_) * Z / fx_;
            float Y = (det.center.y - cy_) * Z / fy_;

            object3d_msgs::msg::Object3D obj;
            obj.point.x = X;
            obj.point.y = Y;
            obj.point.z = Z;
            obj.width_m = (det.box.width * Z) / fx_;
            obj.height_m = (det.box.height * Z) / fy_;
            
            // 使用辅助函数获取类别名
            obj.class_name = detector_->getClassName(det.class_id);
            obj.score = det.confidence;

            array_msg.objects.push_back(obj);
        }

        if (!array_msg.objects.empty()) {
            publisher_->publish(array_msg);
        }
    }

    double fx_, fy_, cx_, cy_;
    double conf_thres_;
    bool show_image_;
    std::unique_ptr<Yolo11Detector> detector_; // 仅持有指针
    
    message_filters::Subscriber<sensor_msgs::msg::Image> color_sub_;
    message_filters::Subscriber<sensor_msgs::msg::Image> depth_sub_;
    typedef message_filters::sync_policies::ApproximateTime<sensor_msgs::msg::Image, sensor_msgs::msg::Image> SyncPolicy;
    std::shared_ptr<message_filters::Synchronizer<SyncPolicy>> sync_;
    rclcpp::Publisher<object3d_msgs::msg::Object3DArray>::SharedPtr publisher_;
    OnSetParametersCallbackHandle::SharedPtr param_callback_handle_;
    std::chrono::steady_clock::time_point start_time_ = std::chrono::steady_clock::now();
    int frame_count_ = 0;
    double fps_ = 0.0;
};

int main(int argc, char** argv) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<YoloDetectNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: src/detect_yolov5_bpu/detect_yolov5_bpu/yolo_detect_node.py
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Image
from cv_bridge import CvBridge
import numpy as np
import cv2
import os
import time
from ament_index_python.packages import get_package_share_directory
from rcl_interfaces.msg import SetParametersResult 
from object3d_msgs.msg import Object3D, Object3DArray
from rclpy.qos import qos_profile_sensor_data
import message_filters
from pathlib import Path

# 导入我们的新模块
from detect_yolov5_bpu.bpu_detect_hobot_dnn import BPU_Detect
import detect_yolov5_bpu.utils as utils

class YoloDetectNode(Node):
    def __init__(self):
        super().__init__('yolo_detect_node')

        # 1. 参数声明 (保持不变)
        self.declare_parameter('camera.fx', 905.5593)
        self.declare_parameter('camera.fy', 905.5208)
        self.declare_parameter('camera.cx', 663.4498)
        self.declare_parameter('camera.cy', 366.7621)
        self.declare_parameter('conf_thres', 0.50)
        self.declare_parameter('nms_thres', 0.45)
        self.declare_parameter('show_image', True)
        self.declare_parameter('model_filename', 'yolov5x_tag_v7.0_detect_640x640_bayese_nv12.bin')
        
        # 获取参数
        self.fx = self.get_parameter('camera.fx').value
        self.fy = self.get_parameter('camera.fy').value
        self.cx = self.get_parameter('camera.cx').value
        self.cy = self.get_parameter('camera.cy').value
        self.show_image_flag = self.get_parameter('show_image').value
        conf_val = self.get_parameter('conf_thres').value
        nms_val = self.get_parameter('nms_thres').value
        model_filename = self.get_parameter('model_filename').value

        self.win_state = {'created': False} 
        if self.show_image_flag:
            dummy_img = np.zeros((720, 1280, 3), dtype=np.uint8)
            cv2.putText(dummy_img, "Loading Model...", (50, 360), cv2.FONT_HERSHEY_SIMPLEX, 1, (255, 255, 255), 2)
            # 立即调用一次显示，创建窗口
            utils.show_window("Detection Result", dummy_img, [], self.win_state)
            # 强制刷新一下事件队列
            cv2.waitKey(1) 

        # 2. 初始化通信
        self.color_sub = message_filters.Subscriber(self, Image, '/camera/realsense_d435i/color/image_raw', qos_profile=qos_profile_sensor_data)
        self.depth_sub = message_filters.Subscriber(self, Image, '/camera/realsense_d435i/aligned_depth_to_color/image_raw', qos_profile=qos_profile_sensor_data)
        self.ts = message_filters.ApproximateTimeSynchronizer([self.color_sub, self.depth_sub], 10, 0.1)
        self.ts.registerCallback(self.sync_callback)
        
        self.publisher_ = self.create_publisher(Object3DArray, 'target_points_array', qos_profile_sensor_data)
        self.bridge = CvBridge()

        # 3. 变量初始化
        self.frame_count = 0
        self.start_time = time.time()
        self.last_log_time = time.time()
        self.fps = 0.0
        # 窗口状态字典 (用于引用传递)
        self.win_state = {'created': False} 

        self.labelname = [
            "person", "bicycle", "car", "motorcycle", "airplane", "bus", "train", "truck", "boat", "traffic light", 
            "fire hydrant", "stop sign", "parking meter", "bench", "bird", "cat", "dog", "horse", "sheep", "cow", 
            "elephant", "bear", "zebra", "giraffe", "backpack", "umbrella", "handbag", "tie", "suitcase", "frisbee", 
            "skis", "snowboard", "sports ball", "kite", "baseball bat", "baseball glove", "skateboard", "surfboard", "tennis racket", "bottle", 
            "wine glass", "cup", "fork", "knife", "spoon", "bowl", "banana", "apple", "sandwich", "orange", 
            "broccoli", "carrot", "hot dog", "pizza", "donut", "cake", "chair", "couch", "potted plant", "bed", 
            "dining table", "toilet", "tv", "laptop", "mouse", "remote", "keyboard", "cell phone", "microwave", "oven", 
            "toaster", "sink", "refrigerator", "book", "clock", "vase", "scissors", "teddy bear", "hair drier", "toothbrush"
        ]

        # 4. 加载模型
        if os.path.isabs(model_filename):
            model_path = model_filename
        else:
            try:
                # 1. 获取包路径并转为 Path 对象
                p = Path(get_package_share_directory('detect_yolov5_bpu'))
                
                # 2. 向上查找直到找到 'install' 目录
                while p.name != 'install' and p != p.parent:
                    p = p.parent
                
                # 3. 拼接 workspace/models 路径
                if p.name == 'install':
                    model_path = str(p.parent / 'models' / model_filename)
                else:
                    self.get_logger().error("Failed to locate 'install' directory.")
                    return
            except Exception as e:
                self.get_logger().error(f"Path error: {e}")
                return

        # 检查与加载
        if not os.path.exists(model_path):
            self.get_logger().error(f"[Error] Model not found: {model_path}")
            return

        self.get_logger().info(f"Loading model: {model_path}")
        self.detector = BPU_Detect(model_path, self.labelname, conf=conf_val, iou=nms_val, is_save=False)
        self.add_on_set_parameters_callback(self.parameter_callback)
        self.get_logger().info("YOLOv5 BPU Node Initialized.")

    def parameter_callback(self, params):
        for param in params:
            if param.name == 'show_image':
                self.show_image_flag = param.value
        return SetParametersResult(successful=True)

    def sync_callback(self, msg_color, msg_depth):
        # 1. FPS 计算
        self.frame_count += 1
        curr_time = time.time()
        if curr_time - self.start_time >= 1.0:
            self.fps = self.frame_count / (curr_time - self.start_time)
            self.frame_count = 0
            self.start_time = curr_time

        # 2. 图像转换
        try:
            cv_image = self.bridge.imgmsg_to_cv2(msg_color, 'bgr8')
            depth_image = self.bridge.imgmsg_to_cv2(msg_depth, '16UC1')
        except Exception as e:
            return

        # 3. 推理 (返回 UnifiedResult 列表)
        results = self.detector.detect(cv_image)

        # 4. 3D 计算与发布
        if self.publisher_.get_subscription_count() > 0:
            array_msg = Object3DArray()
            array_msg.header = msg_color.header

            for res in results:
                # 使用 utils 提取深度
                d_m = utils.get_robust_depth(depth_image, res.center[0], res.center[1])
                
                if d_m <= 0: continue

                # 投影计算
                X = (res.center[0] - self.cx) * d_m / self.fx
                Y = (res.center[1] - self.cy) * d_m / self.fy
                
                obj = Object3D()
                obj.point.x = X
                obj.point.y = Y
                obj.point.z = d_m
                obj.width_m = (res.box[2] * d_m) / self.fx
                obj.height_m = (res.box[3] * d_m) / self.fy
                obj.class_name = res.class_name
                obj.score = res.score
                
                array_msg.objects.append(obj)

            if array_msg.objects:
                self.publisher_.publish(array_msg)

        # 5. 可视化 (使用 utils)
        if self.show_image_flag:
            utils.show_window("Detection Result", cv_image, results, self.win_state, self.fps)
        else:
            # 处理窗口关闭
            if self.win_state['created']:
                cv2.destroyWindow("Detection Result")
                self.win_state['created'] = False
                cv2.waitKey(1)
            
            # 打印日志代替显示
            if curr_time - self.last_log_time >= 5.0:
                self.get_logger().info(f"FPS: {self.fps:.2f}, Objects: {len(results)}")
                self.last_log_time = curr_time

def main(args=None):
    rclpy.init(args=args)
    rclpy.spin(YoloDetectNode())
    rclpy.shutdown()